Add streaming methods to Service infra

richardm-stripe opened the pull request:
r? @dcr-stripe 

https://github.com/stripe/stripe-php/pull/1143 was sufficient for adding Streaming support to `CurlClient`
This PR adds analogous methods onto `AbstractStripeService`, `BaseStripeClient` and such. 

Adding `requestStream` onto `HttpClient\ClientInterface` directly would technically be a breaking change, so I am now exposing separate interfaces for `requestStream` and `request`, and I store two separate references on ApiRequestor.

## Changelog
- Add support for streaming requests on the Services interfaces
  - Add support for `setStreamingHttpClient` and `streamingHttpClient` to `ApiRequestor`
  - Add support for `getStreamingClient` and `requestStream` to `AbstractService`
  - Add support for `requestStream` to `BaseStripeClient`
- `\Stripe\RequestOptions::parse` now clones its input if it is already a `RequestOptions` object, to prevent accidental mutation.

dcr-stripe commented on lib/BaseStripeClient.php:
```diff
@@ -139,6 +139,26 @@ public function request($method, $path, $params, $opts)
+        $baseUrl = $opts->apiBase ?: $this->getApiBase();
+        $requestor = new \Stripe\ApiRequestor($this->apiKeyForRequest($opts), $baseUrl);
+        list($response, $opts->apiKey) = $requestor->requestStream($method, $path, $readBodyChunkCallable, $params, $opts->headers);
+        $opts->discardNonPersistentHeaders();
```
Do we need this line? We have it in request because we then include `$opts` in `$obj` which is returned, but this looks like it would have no impact.

richardm-stripe replied:
Fixed!

dcr-stripe commented on tests/TestHelper.php:
```diff
@@ -95,6 +102,41 @@ function ($method, $absUrl, $headers, $params, $hasFile) {
+     *   Defaults to false.
+     * @param null|string $base base URL (e.g. 'https://api.stripe.com')
+     */
+    protected function expectsRequestStream(
```
Should we add a test that uses this? Maybe on Quotes?

richardm-stripe replied:
Yes, I will rebase [`latest-codegen`](https://github.com/stripe/stripe-php/pull/1152/commits/b4ed515e9d52a124ae02afc5ec6baf0d19d9c0fa) after this is merged, and that PR contains such a test.

Wanted to separate out the infra stuff in this PR vs. the actual quotes-related functionality.

dcr-stripe commented on init.php:
```diff
@@ -18,6 +18,7 @@
 // HttpClient
 require __DIR__ . '/lib/HttpClient/ClientInterface.php';
 require __DIR__ . '/lib/HttpClient/CurlClient.php';
+require __DIR__ . '/lib/HttpClient/StreamingClientInterface.php';
```
I think the order here might matter? You seem to be getting a build break:

```
PHP Fatal error:  Uncaught Error: Interface 'Stripe\HttpClient\StreamingClientInterface' not found in /home/runner/work/stripe-php/stripe-php/lib/HttpClient/CurlClient.php:27
```

As `CurlClient` needs `StreamingClientInterface`, you need to swap these.

richardm-stripe replied:
Yes! Thank you!!

richardm-stripe commented:
r? @dcr

- Added the `clone` we discussed to RequestOptions::parse()
- Addressed code review comments
  - Removed redundant "removeNonPersistentHeaders"
  - Fixed ordering in `init.php`
- Removed trailing commas that were breaking old versions of PHP.

dcr-stripe reviewed (approved):
LGTM! Thanks Richard!
